tools/codegen/gen.py

```diff
@@ -588,7 +588,21 @@ def compute_native_function_declaration(f: NativeFunction) -> List[str]:
         seen.add(n)
         returns_type = native.returns_type(f.func.returns)
         args = native.arguments(f.func)
-        rs.append(f"CAFFE2_API {returns_type} {n}({', '.join(a.str_with_default() for a in args)});")
+
+        def arg_to_str(arg: NativeArgument) -> str:
+            if f.func.is_out_fn():
+                # out overloads don't get default arguments because
+                # defaulted arguments would be before the out argument
+                # in the argument list and that doesn't work.
+                # TODO We should consider if we just want to remove
+                # default arguments from all at::native functions
+                # but that would be a larger change because we need
+                # to change a lot of call sites
```

Comment:
Sorry if I'm being dense but won't only kernels for `c10-full` out ops have the problem described here?  IIUC these are declarations of user-supplied kernels, meaning a port to c10-full would need to take place before this situation arises - until then, the out-first signatures of the existing kernels would necessarily continue to appear here.

If so, we should restrict the default stripping to c10-full out ops, to avoid breaking existing direct calls for no reason.

Orthogonally, minor point but why not just a simple if statement that puts the two different versions of the arg list together? Avoids the nested function and reevaluating the test for every argument.

Reply:
Restricting it to c10 full ops would complicate the landscape because a caller now has to know if it's c10-full or not. Treating all ops the same is more consistent since c10-fullness is an implementation detail that isn't supposed to have API surface. On another note, they're going to be all c10-full soon anyways. With this PR, the overwhelming majority of out ops is already c10-full, so it probably doesn't matter. Note also that at::native isn't an API that we guarantee bc for anyways.

These are declarations of user supplied kernels but default arguments should only occur in the declaration, not in the user supplied definition of the kernel. So this shouldn't have any effect on writers of at::native kernels, only on people who call that API because they don't want to go through the dispatcher.

Reply:
But `NativeFunctions.h` declares user-written kernels, right? 

If so, then kernels for hacky wrapper out ops don't have the issue, since their out args are still at the beginning. For those I think this change will needlessly strip defaults. Knowing whether a kernel is c10-full or not doesn't really apply - it's a matter of whether we leave existing calls undisturbed or not. 

(It's true that moving an op from hacky_wrapper to c10-full will require porting the native kernels and their callers, but I don't see the benefit in churning the callers prior to that.)

Concrete example: here's a kernel from NativeFunctions.h before this change:
```
CAFFE2_API Tensor & cumprod_out(Tensor & out, const Tensor & self, int64_t dim, c10::optional<ScalarType> dtype=c10::nullopt);
```
after:
```
CAFFE2_API Tensor & cumprod_out(Tensor & out, const Tensor & self, int64_t dim, c10::optional<ScalarType> dtype);
```

Note that this is an out op that was hacky wrapper *already*, before this PR, so clearly the default wasn't a problem.

Reply:
ok you convinced me, I'll only remove them for c10-full ops.